RFC-0004-functions-stats.md

```diff
@@ -0,0 +1,420 @@
+    @Description("substring of given length starting at an index")
+    @ScalarFunction
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
+    public static Slice substr(
+            @ScalarPropagateSourceStats(
+                    nullFraction = USE_SOURCE_STATS,
+                    distinctValuesCount = "value(arg_length) <3 ? estimate_ndv(x, value(arg_length), BASIC_LATIN, min(output_row_count, source_stats_ndv())) * (1- source_stats_null_fraction()) : UNKNOWN",
+                    avgRowSize = "value(arg_length)" ) @SqlType("varchar(x)") Slice utf8,
+            @SqlType(StandardTypes.BIGINT) long start,
+            @SqlType(StandardTypes.BIGINT) long length)
+    { //...
+    }
+```
+
+`substr` is at the heart of all `LIKE %X` statements, which is widely used. Without a more powerful framework support it is difficult to support a stats estimation. 
+
+See prior art, for how postgres handles it.
+
+## Metrics
+
+How can we measure the impact of this feature?
+
+1. Existing TPCDS and TPCH queries which use functions or `LIKE %STR` in join condition will benefit from this optimization.
+2. Join order benchmark: Reference [2](#references)
+
+## Other Approaches Considered
+
+Based on the discussion, this may need to be updated with feedback from reviewers.
+
+* __Approach 1__. Trying to find a heuristic based mechanism to estimate functions cost, for example : A function takes a single argument of type `varchar(20)` and
+returns the value of same type i.e. `varchar(20)` and is SCALAR and deterministic, then we can apply heuristic and say that source statistics can be propagated as is.
+
+Example 2: The input and output column do not match, input is `varchar(200)` and output is `varchar(10)`.Then we can say that average row size is changed.
+
+__Pros:__
+
+* Minimum impact to the user and codebase. Since this feature can be 
+enabled/disabled using a session flag : `scalar_function_stats_propagation_enabled`
+
+__Cons:__
+* Problem with heuristic based approach is unknown nature of functions, and thus it can be difficult to come up with heuristics that are always accurate.
+
+* __Approach 2__. Reference [1. Monsoon](#references). discusses various stochastic approaches to solve the same problem, a user provided stats calculator will complement such
+approaches as one could compute via statistical process for those functions user provided stats computation is unavailable. Sampling and other stochastic processes are beyond the
+scope for this RFC.
+
+## Adoption Plan
+
+- What impact (if any) will there be on existing users? Are there any new session parameters, configurations, SPI updates, client API updates, or SQL grammar?
+
+
+Functions without this annotation will not be affect, to remain backwards compatible those functions without this Annotation will run as is. Also, a new session 
+config will be introduced i.e. `optimizer.scalar-function-stats-propagation-enabled=false`, by setting this flag this feature can be turned off.
```

Comment:
`optimizer.scalar-function-stats-propagation-enabled=false` as currently written is in the form of a "FeatureConfig" (most session properties have only `_` in their name, `-` are not allowed).

```suggestion
config will be introduced i.e. `optimizer_scalar_function_stats_propagation_enabled`, by setting this flag to `false` this feature can be turned off.
```

Reply:
I was intending feature config. Session config is specified at https://github.com/prestodb/rfcs/pull/14/files#diff-563baec9193d2d661febe4c2fd74c485dfcb16a9027ee8284dea76be7c1a1915R473 .

Reply:
Then in the previous line we should call it feature config and not a session config